### PR TITLE
[GCP] Adopt new provisioner to stop/down clusters

### LINKED ISSUE #1983: Update GCP node provider to terminate nodes in parallel
User reported that `sky down` on multinode GCP clusters is too slow (either during failover or an actual down call). This is because the node provider impl terminates nodes serially.

We could merge in https://github.com/ray-project/ray/pull/34455 and test.

### LINKED ISSUE #2089: [Core] Resource leakage for sky down if a multi-node cluster is partially stopped
If a multi-node cluster is partially stopped (during autostop or manually stop the worker node), i.e. the cluster is in INIT state, our `backend.teardown` function will fail to terminate the stopped nodes, as we are using `ray down` to terminate the cluster, and ray ignores the stopped nodes.
Related code path: https://github.com/skypilot-org/skypilot/blob/b9626015fd4cd01ab29d4354ba38c5ff17884e3d/sky/backends/cloud_vm_ray_backend.py#L3367-L3369

To reproduce:
1. `sky launch -c min --cloud gcp --num-nodes 2 --cpus 2`
2. Manually stop the worker node.
3. `sky down min`

After that, the head node will be terminated, but the worker node is still in STOPPED state.

One possible solution is to only use our own cloud cli-based termination (should be careful: may need to kill the head node first to avoid ray autoscaler restarting some of the worker nodes, causing leakage).

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR adopts the new provisioner refactoring from #2121. 

Fixes #2089, to avoid the nodes to be leaked when the cluster is partially down.
Also Fixes #1983, as we will now send the requests first and wait the completion of together (using multiple threads in https://github.com/ray-project/ray/pull/34455 is buggy, refer to https://github.com/skypilot-org/skypilot/pull/2160#issue-1783260816)

Previously:
Stopping 8 nodes takes: 168s

Now:
Stopping 8 nodes takes: 40s

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch -c test-gcp-mn --num-nodes 8 --cpus 2+ --cloud gcp; sky stop -y test-gcp-mn; sky down -y test-gcp-mn`
  - [x] `sky launch -c test-gcp-mn --num-nodes 8 --cpus 2+  --cloud gcp`; manually stop some of the nodes; `sky down -y test-gcp-mn`
  - [x] `sky launch -c test-gcp-tpu --cpus 2+ --gpus tpu-v2-8; sky stop test-gcp-tpu-mn; sky down -y test-gcp-tpu-mn`
  - [x] `sky launch -c test-tpu-vm test.yaml; sky stop test-tpu-vm; sky start test-tpu-vm; sky down test-tpu-vm`
    ```
    resources:
        accelerators: tpu-v2-8
        accelerator_args:
            tpu_vm: true
    ```
  - [x] `sky launch -c test-tpu-vm test.yaml`; wait for the tpu vm to be preempted; `sky status -r test-tpu-vm` correctly terminate the tpu vm on the cloud.
- [x] All smoke tests: `pytest tests/test_smoke.py` (except for `test_cancel_gcp`, `test_distributed_tf` due to the issue of the tensorflow code for resnet training #2207)
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/provision/gcp/instance_utils.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+            cache_discovery=False,
+            discoveryServiceUrl='https://tpu.googleapis.com/$discovery/rest')
+        operation = tpu.projects().locations().nodes().stop(
+            name=instance).execute()
+        return operation
+
+    @classmethod
+    def terminate(cls, project_id: str, zone: str, instance: str) -> dict:
+        """Terminate a TPU node."""
+        tpu = gcp.build(
```
Comment (suquark) [inline sky/provision/gcp/instance_utils.py:295]: reuse this gcp build as a class method?

File: sky/provision/gcp/instance.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, cluster_name
+    assert provider_config is not None, cluster_name
+    zone = provider_config['availability_zone']
+    project_id = provider_config['project_id']
+    name_filter = {TAG_RAY_CLUSTER_NAME: cluster_name}
+    handlers: List[Type[instance_utils.GCPInstance]] = [
+        instance_utils.GCPComputeInstance
+    ]
+    use_tpu_vms = provider_config.get('_has_tpus', False)
```
Comment (suquark) [inline sky/provision/gcp/instance.py:138]: move this near the start of the function like other fields

File: sky/provision/gcp/instance.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+def terminate_instances(
+    cluster_name: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    included_instances: Optional[List[str]] = None,
+    excluded_instances: Optional[List[str]] = None,
+) -> None:
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, cluster_name
+    assert provider_config is not None, cluster_name
```
Comment (suquark) [inline sky/provision/gcp/instance.py:131]: duplicated assertion?

File: sky/backends/cloud_vm_ray_backend.py [hunk 4]
```diff
@@ -71,7 +71,7 @@
 _NODES_LAUNCHING_PROGRESS_TIMEOUT = {
     clouds.AWS: 90,
     clouds.Azure: 90,
-    clouds.GCP: 120,
+    clouds.GCP: 240,
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:74]: This is a known issue that the `120` seconds is not enough for launching multiple nodes for GCP. We need to increase this to make sure `sky launch --cloud gcp --num-nodes 8` to work.

File: sky/provision/gcp/instance_utils.py (hunk 1 above)
Comment (Michaelvll) [reply sky/provision/gcp/instance_utils.py]: Good point! Moved to a function now.

**Review by suquark — APPROVED** [review]
LGTM

**Comment by Michaelvll** [comment]
Thanks for reviewing this @suquark! I am now testing it with `pytest tests/test_smoke.py`

**Pull request merged**